src/Microsoft.AspNet.SignalR.ServiceBus3/ServiceBusConnection.cs

```diff
@@ -0,0 +1,340 @@
+                }
+
+                // Create a receiver to get messages
+                string subscriptionEntityPath = SubscriptionClient.FormatSubscriptionPath(topicName, subscriptionName);
+                MessageReceiver receiver = _factory.CreateMessageReceiver(subscriptionEntityPath, ReceiveMode.ReceiveAndDelete);
+
+                _trace.TraceInformation("Creation of a message receive for subscription entity path {0} in the service bus completed successfully.", subscriptionEntityPath);
+
+                connectionContext.SetSubscriptionContext(new SubscriptionContext(topicName, subscriptionName, receiver), topicIndex);
+
+                var receiverContext = new ReceiverContext(topicIndex, receiver, connectionContext);
+
+                var task = Task.Run(() => ProcessMessages(receiverContext));
+                _processMessagesTasks.GetOrAdd(task, false);
+                task.ContinueWith(t =>
+                    {
+                        bool value;
+                        _processMessagesTasks.TryRemove(task, out value);
+                    });
+
+                // Open the stream
+                connectionContext.OpenStream(topicIndex);
+            }
+        }
+
+        [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "We retry to create the topics on exceptions")]
+        private void Retry(Action action)
+        {
+            string errorMessage = "Failed to create service bus subscription or topic : {0}";
+            while (true)
+            {
+                try
+                {
+                    action();
+                    break;
+                }
+                catch (UnauthorizedAccessException ex)
+                {
+                    _trace.TraceError(errorMessage, ex.Message);
+                    break;
+                }
+                catch (MessagingException ex)
+                {
+                    _trace.TraceError(errorMessage, ex.Message);
+                    if (ex.IsTransient)
+                    {
+                        Thread.Sleep(RetryDelay);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    _trace.TraceError(errorMessage, ex.Message);
+                    Thread.Sleep(RetryDelay);
+                }
+            }
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                // Close the factory
+                if (_factory != null)
+                {
+                    _factory.Close();
+                }
+
+                _cts.Cancel();
+                Task.WaitAll(_processMessagesTasks.Keys.ToArray(), DefaultReadTimeout);
```

Comment:
This is the longest timeout we are using in ServiceBusConnection so this should give a chance to finish all the tasks cleanly before giving up.